### PR TITLE
New README

### PR DESCRIPTION
**Pull request by msaroufim**
This a restructuring of our main README.md with the main goals of having
1. Clearer focus on our training, inference finetuning story
2. Clarifying our quantization terminology
3. Clearer benchmarks and API usage

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,135 +92,49 @@ This repository is currently under heavy development - if you have suggestions o
...
+pip install --pre torchao-nightly --index-url https://download.pytorch.org/whl/nightly/cu121 # full options are cpu/cu118/cu121/cu124
 ```
 
-From source
+## Community Contributions
+
+* [jeromeku](https://github.com/jeromeku) has implemented
+    * [GaLore](torchao/prototype/galore/) a drop for the Adam Optimizer that allows you to finetune llama 7b on a single 4090 card with up to 70% speedups relative to eager PyTorch
+    * [DoRA](torchao/prototype/dora) a newer replacement for QLoRA with more promising convergence characteristics
+    * [Fused int4/fp16 Quant Matmul](torchao/prototype/hqq) which is particularly useful for compute bound kernels showing 4x speedups over tinygemm for larger batch sizes such as 512
```
Comment (gau-nernst) [inline README.md:108]: tinygemm is mentioned here. Should we link it to a doc section talking about tinygemm, if it exists?

Comment (msaroufim) [reply README.md:108]: Yeah tinygemm is a great idea, will add

File: README.md [hunk 2]
```diff
@@ -2,147 +2,142 @@
 
 [![](https://dcbadge.vercel.app/api/server/cudamode?style=flat)](https://discord.gg/cudamode)
 
-This repository is currently under heavy development - if you have suggestions on the API or use-cases you'd like to be covered, please open an [issue](https://github.com/pytorch/ao/issues)
 
 ## Introduction
-`torchao` is a PyTorch library for quantization and sparsity.
+
+torchao is a library which makes it easy to integrate and create high performance kernels with custom data types and layouts with up to
```
Comment (supriyar) [inline README.md:8]: create, integrate and compare?

File: README.md [hunk 3]
```diff
@@ -2,147 +2,142 @@
...
+
+All with no intrusive code changes and minimal accuracy degradation.
+
+## Benchmarks
+
+### Inference
+
+#### Without intrusive code changes
+
+Quantizing your models is a 1 liner that should work on any model with `nn.Linear` including your favorite HuggingFace model. You can find a more comprehensive usage example [here](torchao/quantization/)
```
[the remark this answers is not in the document]
Comment (msaroufim) [reply README.md:20]: It's outdated now because of the new quantization API changes but this just works

https://github.com/pytorch/ao/blob/main/scripts/hf_eval.py

File: README.md [hunk 4]
```diff
@@ -2,147 +2,142 @@
...
+|             | int8wo             | 12.204              |  147.03       | 973.54                  | 6.62             | 8.95            |
+|             | int4wo-64          | 12.843              |  199.81       | 746.45                  | 3.74             | 4.75            |
+|             | int4wo-64-GPTQ     | 12.489              |  199.81       | 746.45                  | 3.74             | 4.75            |
+| Llama-3-8B  | Base (bfloat16)    | N/A                 |  94.91        | 1424.58                 | 15.01            | 16.43           |
+|             | int8dq             | N/A                 |  8.41         | 63.23                   | 7.52             | 9.24            |
+|             | int8wo             | N/A                 |  136.75       | 1028.38                 | 7.52             | 10.42           |
+|             | int4wo-64          | N/A                 |  179.41       | 757.45                  | 4.22             | 6.88            |
+
+note: Int8 dynamic quantization works best on compute bound models like [SAM](https://github.com/pytorch-labs/segment-anything-fast) whereas Llama with batchsize=1 tends to be memory bound, thus the rather low performance.
```
Comment (supriyar) [inline README.md:43]: should we add a note here on why perplexity results are NA for llama3?

File: README.md (hunk 2 above)
Comment (msaroufim) [reply README.md:8]: Not sure how I feel about compare being in the first line, if the intent is to say carefully benchmarked then I feel that's table stakes for a library about performance and not worth explicitly mentioning

File: README.md [hunk 5]
```diff
@@ -2,147 +2,142 @@
...
+
+And a quick crash course on inference quantization to help parse the above table. Int4 quantization is an ambiguous term because there's the dtype in which a layer is represented and then the dtype in which the computation is done. For example, if you're using Weight-Only (wo) int4 quantization that means that the layer will be upcasted to a larger dtype like fp16 so an int4 matrix multiplication is defined as `F.linear(input, weight.to(input.dtype))` whereas if it's possible to perform the computation using the smaller dtype directly pending support by a hardware vendor then that means you can perform `F.linear(input, weight)` directly and this is what we refer to as Dynamic-Quantization (dq). Naive quantization algorithms are also notoriously sensitive to outliers so we also typically set a group size that applies a scale factor per group of 64 elements in the case of `int4wo64`.
```
Comment (supriyar) [inline README.md:45]: > whereas if it's possible to perform the computation using the smaller dtype directly pending support by a hardware vendor then that means you can perform `F.linear(input, weight)` directly and this is what we refer to as Dynamic-Quantization (dq)

dynamic quantization is used for the activations specifically. So activations are quantized dynamically (bf16->int8) and weight is quantized to int4 or int8 and computation happens in lower precision.

Theoretically you could have dynamic quantization where the weights might also get upcasted (but this doesn't happen in practice afaik).

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (msaroufim) [reply README.md:20]: cc @jerryzh168 @HDCharles

File: README.md [hunk 6]
```diff
@@ -2,147 +2,142 @@
...
+
+In some cases we rewrote popular GenAI models to be significantly faster in native PyTorch as in no C++/CUDA to achieve at the time SOTA inference performance. These involve more intrusive code changes.
+
+* 8x speedups for Image segmentation models with [sam-fast](https://pytorch.org/blog/accelerating-generative-ai)
+* 10x speedups for Language models with [gpt-fast](https://pytorch.org/blog/accelerating-generative-ai-2)
+* 3x speedup for Diffusion models with [sd-fast](https://pytorch.org/blog/accelerating-generative-ai-3)
+
+### Training
+
+We've added support for semi-structured 2:4 sparsity with over 30% speedups on ViT-L
```
Comment (supriyar) [inline README.md:58]: to clarify, the speedup here is not for the entire model. I believe its specifically for the MLP blocks in ViT. cc @jcaip

File: README.md [hunk 7]
```diff
@@ -2,147 +2,142 @@
...
+The code change is a 1 liner with the full example available [here](torchao/sparsity/training/)
+
+
+```python
+swap_linear_with_semi_sparse_linear(model, {"seq.0": SemiSparseLinear})
+```
+
+For VIT-L MLP shapes on a NVIDIA A100 we see the following results:
+
+|                     |   act24   |   dense   |   w24    | s24_inp_sparsify24 | s24_inp_clone |
```
Comment (supriyar) [inline README.md:69]: discussed this with @jcaip offline but would be good to make the columns here more clear.

File: README.md [hunk 8]
```diff
@@ -2,147 +2,142 @@
...
+
+Times are in microseconds (us).
+
+## Newer dtypes
+
+[MX](https://github.com/pytorch/ao/blob/main/torchao/prototype/mx_formats) implementing training and inference support with tensors using the [OCP MX spec](https://www.opencompute.org/documents/ocp-microscaling-formats-mx-v1-0-spec-final-pdf) data types, which can be described as groupwise scaled float8/float6/float4/int8, with the scales being constrained to powers of two. This work is prototype as the hardware support is not available yet.
+
+[nf4](https://github.com/pytorch/ao/blob/main/torchao/dtypes/nf4tensor.py) which was used to [implement QLoRA](https://github.com/pytorch/torchtune/blob/main/docs/source/tutorials/qlora_finetune.rst) one of the most popular finetuning algorithms without writing custom Triton or CUDA code. Accessible talk [here](https://x.com/HamelHusain/status/1800315287574847701)
+
+[tinygemm](https://github.com/pytorch/ao/blob/cb3bd8c674f2123af232a0231b5e38ddafa756a8/torchao/dtypes/aqt.py#L526) we make heavy use of `torch.ops.aten._weight_int4pack_mm` to bitpack into a layout optimized for tensor cores
```
Comment (supriyar) [inline README.md:81]: why is this under newer dtypes and not performant kernels or something?

File: README.md (hunk 7 above)
Comment (msaroufim) [reply README.md:69]: @jcaip please send me a clearer table when you have it - keep in mind I want to make some claim about end to end performance

File: README.md (hunk 6 above)
Comment (msaroufim) [reply README.md:58]: can we put together anything from an end to end perspective

File: README.md (hunk 2 above)
Comment (supriyar) [reply README.md:8]: when I read this sentence, to me torchao comes across as a kernel library with custom types. But I think what we want to say is something like "torchao is a library to create and integrate high-performance custom data types, layouts and kernels into their PyTorch workflows with up to.." wdyt?

File: README.md (hunk 7 above)
Comment (jcaip) [reply README.md:71]: Yes, I'm making some changes to the benchmarking script for e2e ViT. Will send something by EOD.

File: README.md (hunk 5 above)
Comment (msaroufim) [reply README.md]: Is it clearer now?

File: README.md (hunk 2 above)
Comment (msaroufim) [reply README.md:8]: Yeah that's much clearer

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (msaroufim) [reply README.md]: Ok just linked to the existing script for now

File: README.md [hunk 9]
```diff
@@ -2,147 +2,142 @@
...
+
+For VIT-L MLP shapes on a NVIDIA A100 we see the following results:
+
+|                     |   act24   |   dense   |   w24    | s24_inp_sparsify24 | s24_inp_clone |
+|---------------------|-----------|-----------|----------|--------------------|---------------|
+| f16 (44160,1024,4096,1024) |  11881.0  |  11534.3  |  9204.7  |        255.1        |      125.8    |
+
+Times are in microseconds (us).
+
+## Newer dtypes
```
Comment (supriyar) [inline README.md:77]: should we also mention the fp6 dtype here?

**Review by supriyar — APPROVED** [review]
looks great! 🚀 

Ready to ship once we add the latest sparsity numbers.

File: README.md [hunk 10]
```diff
@@ -2,147 +2,142 @@
...
+|             | int8wo             | 12.204              |  147.03       | 973.54                  | 6.62             | 8.95            |
+|             | int4wo-64          | 12.843              |  199.81       | 746.45                  | 3.74             | 4.75            |
+|             | int4wo-64-GPTQ     | 12.489              |  199.81       | 746.45                  | 3.74             | 4.75            |
+| Llama-3-8B  | Base (bfloat16)    |                  |  94.91        | 1424.58                 | 15.01            | 16.43           |
+|             | int8dq             |                  |  8.41         | 63.23                   | 7.52             | 9.24            |
+|             | int8wo             |                  |  136.75       | 1028.38                 | 7.52             | 10.42           |
+|             | int4wo-64          |                  |  179.41       | 757.45                  | 4.22             | 6.88            |
+
+note: Int8 dynamic quantization works best on compute bound models like [SAM](https://github.com/pytorch-labs/segment-anything-fast) whereas Llama with batchsize=1 tends to be memory bound, thus the rather low performance.
```
Comment (vkuzo) [inline README.md:43]: I thought it was more about compute bound vs memory bound, and less about SAM vs LLaMa

File: README.md (hunk 9 above)
Comment (msaroufim) [reply README.md:77]: yep

File: README.md (hunk 10 above)
Comment (msaroufim) [reply README.md:43]: indeed that's intent of the phrase but seems like it's not as clear as I'd thought

File: README.md (hunk 7 above)
Comment (msaroufim) [reply README.md:71]: Btw depending on what time you're done feel free to just push changes to my PR directly and merge since I'll be unavailable tomorrow

File: README.md (hunk 10 above)
Comment (msaroufim) [reply README.md]: better?

File: README.md [hunk 11]
```diff
@@ -2,147 +2,142 @@
...
+|             | int4wo-64          | 12.843              |  199.81       | 746.45                  | 3.74             | 4.75            |
+|             | int4wo-64-GPTQ     | 12.489              |  199.81       | 746.45                  | 3.74             | 4.75            |
+| Llama-3-8B  | Base (bfloat16)    |                  |  94.91        | 1424.58                 | 15.01            | 16.43           |
+|             | int8dq             |                  |  8.41         | 63.23                   | 7.52             | 9.24            |
+|             | int8wo             |                  |  136.75       | 1028.38                 | 7.52             | 10.42           |
+|             | int4wo-64          |                  |  179.41       | 757.45                  | 4.22             | 6.88            |
+
+note: Int8 dynamic quantization works best on compute bound as opposed to memory bound models. Some relatable examples might be [SAM](https://github.com/pytorch-labs/segment-anything-fast) which is memory bound vs Llama at batchsize=1 which is memory bound.
```
Comment (supriyar) [inline README.md:43]: > [SAM](https://github.com/pytorch-labs/segment-anything-fast) which is memory bound

typo -> should be "compute bound"
